### PR TITLE
feat: unify extended Python IPC & consolidate agent startup logic

### PR DESCRIPTION
**Pull request by isidentical**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30
* #29
* #28
* #27
* #25
* #23
* #22
* #21
* #20
* __->__ #19

### CODE REVIEW

File: src/isolate/backends/connections/agent_startup.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    python_path = os.getenv("PYTHONPATH")
+    if python_path is None:
+        return None
+
+    # TODO: The order here is the same as the one that is used for generating the
+    # PYTHONPATH. The only problem that might occur is that, on a chain with
+    # 3 ore more nodes (A, B, C), if X is installed as an editable package to
+    # B and a normal package to C, then C might actually take precedence. This
+    # will need to be fixed once we are dealing with more than 2 nodes and editable
+    # packages.
```
Comment (mederka) [inline src/isolate/backends/connections/agent_startup.py:28]: 👍

File: src/isolate/backends/connections/ipc/_base.py [hunk 2]
```diff
@@ -176,50 +195,94 @@ def poll_until_result(
...
-        the given environment_path."""
+        """Start the agent process with the Python binary available inside the
+        bound environment."""
 
         python_executable = get_executable_path(self.environment_path, "python")
         with logged_io(
-            partial(self._parse_agent_and_log, level=LogLevel.STDOUT),
-            partial(self._parse_agent_and_log, level=LogLevel.STDERR),
+            partial(self.handle_agent_log, level=LogLevel.STDOUT),
+            partial(self.handle_agent_log, level=LogLevel.STDERR),
```
Comment (mederka) [inline src/isolate/backends/connections/ipc/_base.py:232]: 👍

File: src/isolate/backends/connections/ipc/_base.py [hunk 3]
```diff
@@ -176,50 +195,94 @@ def poll_until_result(
...
             )
 
-    def _get_python_env(self):
+    def get_env_vars(self) -> Dict[str, str]:
+        """Return the environment variables to run the agent process with. By default
+        PYTHONUNBUFFERED is set to 1 to ensure the prints to stdout/stderr are reflect
+        immediately (so that we can seamlessly transfer logs)."""
+
+        custom_vars = {}
+        custom_vars["PYTHONUNBUFFERED"] = "1"
```
Comment (mederka) [inline src/isolate/backends/connections/ipc/_base.py:248]: ATL: how is this var related to printing std?

**Review by mederka — APPROVED** [review]
Great job, @isidentical!

File: src/isolate/backends/connections/ipc/_base.py (hunk 3 above)
Comment (isidentical) [reply src/isolate/backends/connections/ipc/_base.py:248]: The way it currently works is that, the controller process spawns threads to watch the stdout/stderr of the newly created agent process. By default, buffering is enabled for python, so there will be delay between the threads in the controller seeing the actual stdout log and the user's script printing it. So we set `PYTHONUNBUFFERED` to ensure that everything is continuously streamed and not buffered from the agent.
